controller/composite/controller.go

```diff
@@ -580,17 +580,23 @@ func (pc *parentController) claimChildren(parent *unstructured.Unstructured) (co
 }
 
 func (pc *parentController) updateParentStatus(parent *unstructured.Unstructured, status map[string]interface{}) (*unstructured.Unstructured, error) {
+	// Inject ObservedGeneration before comparing with old status,
+	// so we're comparing against the final form we desire.
+	if status == nil {
+		status = make(map[string]interface{})
+	}
+	status["observedGeneration"] = parent.GetGeneration()
+
 	// Overwrite .status field of parent object without touching other parts.
 	// We can't use Patch() because we need to ensure that the UID matches.
 	return pc.parentClient.Namespace(parent.GetNamespace()).AtomicStatusUpdate(parent, func(obj *unstructured.Unstructured) bool {
-		oldStatus := k8s.GetNestedField(obj.UnstructuredContent(), "status")
+		oldStatus := obj.UnstructuredContent()["status"]
 		if reflect.DeepEqual(oldStatus, status) {
 			// Nothing to do.
 			return false
 		}
 
-		k8s.SetNestedField(obj.UnstructuredContent(), status, "status")
-		k8s.SetNestedField(obj.UnstructuredContent(), parent.GetGeneration(), "status", "observedGeneration")
```

Comment:
`.status.observedGeneration` is not needed to be set explicitly with the change because `obj.UnstructuredContent()["status"]` will set all nested status fields accordingly. Correct me if I am wrong.

Reply:
That's correct. It would have also worked with SetNestedField, since both will fully replace the entire `.status` sub-tree of the object. I switched it to map notation to make it more clear that we're working with maps. I think in this form, the original bug might have been less likely since we're used to thinking about the possibility of nil maps.